### PR TITLE
Metadata based filtering for test

### PR DESCRIPTION
**Pull request by yogeshnikam671**
**What**:
Add ability to filter tests based on the metadata of a certain test.
Add --filter and --filterNot argument in TestCommand to use these filters.
The filters follow a specific syntax which can be understood using `specmatic  test --help` command.

**Why**:
The existing --filterName and --filterNotName were serving the purpose of filtering tests to an extent but potentially that algorithm is not scalable when it comes to evolving the filtering mechanism and making it more scalable.
Hence, this metadata based filtering is introduced. This is a first cut of it.

**Checklist**:

- [ ] Documentation  (Pending, will do it in parallel)
- [x] Tests
- [ ] Sonar Quality Gate

### CODE REVIEW

File: core/src/main/kotlin/io/specmatic/core/HttpRequestPattern.kt [hunk 1]
```diff
@@ -30,6 +30,10 @@ data class HttpRequestPattern(
     val multiPartFormDataPattern: List<MultiPartFormDataPattern> = emptyList(),
     val securitySchemes: List<OpenAPISecurityScheme> = listOf(NoSecurityScheme())
 ) {
+    fun getHeaderKeys() = headersPattern.pattern.keys
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/core/HttpRequestPattern.kt]: Instead of digging into headers for data, wrap this in something like `headersPattern.headerNames`

File: core/src/main/kotlin/io/specmatic/core/HttpRequestPattern.kt [hunk 2]
```diff
@@ -30,6 +30,10 @@ data class HttpRequestPattern(
...
+
+    fun getQueryParamKeys() = httpQueryParamPattern.queryPatterns.keys
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/core/HttpRequestPattern.kt]: Similar to the previous comment, httpQueryParamPattern.queryParamNames

File: core/src/main/kotlin/io/specmatic/core/filters/ScenarioFilterTags.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+import io.specmatic.core.HEADERS_BREADCRUMB
+import io.specmatic.core.METHOD_BREAD_CRUMB
+import io.specmatic.core.PATH_BREAD_CRUMB
+import io.specmatic.core.QUERY_PARAMS_BREADCRUMB
+
+enum class ScenarioFilterTags(val key: String) {
+    METHOD(METHOD_BREAD_CRUMB),
+    PATH(PATH_BREAD_CRUMB),
+    STATUS_CODE("STATUS-CODE"),
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/core/filters/ScenarioFilterTags.kt]: Can you confirm that we do not already have a word for this, by checking in the status matching code HttpResponsePattern.

File: application/src/main/kotlin/application/TestCommand.kt [hunk 4]
```diff
@@ -82,6 +84,48 @@ class TestCommand : Callable<Unit> {
...
+For example:
+--filter="METHOD=GET,POST" --filter="PATH=/users"
+           """
+        ],
+        required = false
+    )
+    var filter: List<String> = emptyList()
+
+    @Option(
+        names= ["--filterNot"],
```
Comment (joelrosario) [inline application/src/main/kotlin/application/TestCommand.kt]: this should be `--filter-not`, in keeping with `--filter-name` and `--filter-not-name`

File: core/src/main/kotlin/io/specmatic/core/utilities/Utilities.kt [hunk 5]
```diff
@@ -334,7 +334,9 @@ fun saveJsonFile(jsonString: String, path: String, fileName: String) {
 }
 
 fun readEnvVarOrProperty(envVarName: String, propertyName: String): String? {
-    return System.getenv(envVarName) ?: System.getProperty(propertyName)
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/core/utilities/Utilities.kt]: What's the reason for this change?

File: core/src/main/kotlin/io/specmatic/test/ContractTest.kt [hunk 6]
```diff
@@ -9,6 +9,7 @@ interface ResponseValidator {
 }
 
 interface ContractTest {
+    val scenario: Scenario
```
Comment (joelrosario) [inline core/src/main/kotlin/io/specmatic/test/ContractTest.kt]: Seems like the reason to expose this is because we need the scenario metadata and testDescription elsewhere. How about just exposing those methods in ContractTest, so that no-one has to dig in and chain method calls? Then you could remove the `scenario` property.

**Review by joelrosario — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/io/specmatic/core/utilities/Utilities.kt (hunk 5 above)
Comment (yogeshnikam671) [reply core/src/main/kotlin/io/specmatic/core/utilities/Utilities.kt]: On retrospect I don't need this change. I'll revert. I had added this because I thought if I set env var as empty string, it won't work properly unless I return null from this function. That was an incorrect assumption.

File: core/src/main/kotlin/io/specmatic/test/ContractTest.kt (hunk 6 above)
Comment (yogeshnikam671) [reply core/src/main/kotlin/io/specmatic/test/ContractTest.kt]: This is a very good idea. I'll do that.

File: core/src/main/kotlin/io/specmatic/core/utilities/Utilities.kt (hunk 5 above)
Comment (yogeshnikam671) [reply core/src/main/kotlin/io/specmatic/core/utilities/Utilities.kt]: Addressed.

File: core/src/main/kotlin/io/specmatic/test/ContractTest.kt (hunk 6 above)
Comment (yogeshnikam671) [reply core/src/main/kotlin/io/specmatic/test/ContractTest.kt]: Addressed.

File: application/src/main/kotlin/application/TestCommand.kt (hunk 4 above)
Comment (yogeshnikam671) [reply application/src/main/kotlin/application/TestCommand.kt]: Addressed.

File: core/src/main/kotlin/io/specmatic/core/filters/ScenarioFilterTags.kt (hunk 3 above)
Comment (yogeshnikam671) [reply core/src/main/kotlin/io/specmatic/core/filters/ScenarioFilterTags.kt]: The breadCrumb just says STATUS, shall I change it to STATUS-CODE in breadcrumb in HttpResponsePattern and then use it here?

File: core/src/main/kotlin/io/specmatic/core/HttpRequestPattern.kt (hunk 1 above)
Comment (yogeshnikam671) [reply core/src/main/kotlin/io/specmatic/core/HttpRequestPattern.kt]: Addressed.

File: core/src/main/kotlin/io/specmatic/core/HttpRequestPattern.kt (hunk 2 above)
Comment (yogeshnikam671) [reply core/src/main/kotlin/io/specmatic/core/HttpRequestPattern.kt]: Addressed.

File: core/src/main/kotlin/io/specmatic/core/filters/ScenarioFilterTags.kt (hunk 3 above)
Comment (joelrosario) [reply core/src/main/kotlin/io/specmatic/core/filters/ScenarioFilterTags.kt]: Let this be STATUS for now, but we should change it to STATUS-CODE later.

**Review by joelrosario — APPROVED** [review]
